Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,296 @@
+
+  // This function expects the mutex to be already locked.
+  void DoPop(T* const out, const Iterator where) {
+    FIREBASE_ASSERT_MESSAGE(!scheduled_.empty(),
+                            "Trying to pop an entry from an empty queue.");
+
+    if (out) {
+      *out = std::move(where->value);
+    }
+    scheduled_.erase(where);
+    cv_.notify_one();
+  }
+
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+  Container scheduled_;
+};
+
+class AsyncQueue;
+
+// A non-owning handle to an operation scheduled in the future, allowing to
+// cancel the operation.
+class DelayedOperation {
+ public:
+  // If the operation has not been run yet, cancels the operation. Otherwise,
+  // it's a no-op.
+  void Cancel();
+
+ private:
+  using Id = unsigned int;
+
+  // Don't allow callers to create their own `DelayedOperation`s.
+  friend class AsyncQueue;
+  DelayedOperation(AsyncQueue* const queue, const Id id)
+      : queue_{queue}, id_{id} {
+  }
+
+  AsyncQueue* const queue_ = nullptr;
+  const Id id_ = 0;
+};
+
+// A serial queue that executes provided operations on a dedicated background
+// thread.
+//
+// Operations may be scheduled for immediate or delayed execution. Operations
+// scheduled for the exact same time are FIFO ordered. Immediate operations
+// always come before delayed operations.
+//
+// The operations are executed sequentially; only a single operation is executed
+// at any given time.
+//
+// Delayed operations may be canceled if they have not already been run.
+class AsyncQueue {
+ public:
+  using Operation = std::function<void()>;
+  using Milliseconds = std::chrono::milliseconds;
+
+ private:
+  using TimePoint = Schedule<Operation, Milliseconds>::TimePoint;
+  // To allow canceling operations, each scheduled operation is assigned
+  // a monotonically increasing identifier.
+  using Id = DelayedOperation::Id;
+
+ public:
+  AsyncQueue();
+  ~AsyncQueue();
+
+  // Enqueues the `operation` for immediate execution on the background thread.
+  void Enqueue(Operation&& operation);
+  // Enqueues the `operation` for execution on the background thread once the
+  // `delay` from now (according to the system clock) has passed. Returns
+  // a handle which allows to cancel the delayed operation.
+  //
+  // `delay` must be non-negative; use `Enqueue` to schedule operations for
+  // immediate execution.
+  DelayedOperation EnqueueAfterDelay(Milliseconds delay, Operation&& operation);
+
+ private:
+  friend class DelayedOperation;  // For access to `TryCancel`
+  // If the operation hasn't yet been run, it will be removed from the queue.
+  // Otherwise, this function is a no-op.
+  void TryCancel(const DelayedOperation& operation);
+
+  Id DoEnqueue(Operation&& operation, TimePoint when);
+
+  void PollingThread();
+  void UnblockQueue();
+  unsigned int NextId();
+
+  // As a convention, assign the epoch time to all operations scheduled for
+  // immediate execution. Note that it means that an immediate operation is
```

Comment:
I think this is equivalent to existing Objective-C behavior.

Reply:
Revisiting this, I am less sure. Consider:
```cpp
void print_next_number();
void print_foo();
void foo(AsyncQueue& q) {
  q.EnqueueWithDelay(Milliseconds(1), print_foo);
  while (true)
    q.Enqueue(print_next_number);
}
```

If an immediate operation gets assigned to the time it was enqueued, then a possible output would be:
```
1
2
3
4
5
foo
6
7
```

OTOH, if immediate operation always takes the priority over any delayed operation, then it's possible that "foo" never gets printed.